core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -1413,55 +1412,108 @@ final class Stream[+F[_], +O] private[fs2] (private[fs2] val underlying: Pull[F,
       n: Int,
       timeout: FiniteDuration
   )(implicit F: Temporal[F2]): Stream[F2, Chunk[O]] =
-    this
-      .covary[F2]
-      .pull
-      .timed { timedPull =>
-        def resize(c: Chunk[O], s: Pull[F2, Chunk[O], Unit]): (Pull[F2, Chunk[O], Unit], Chunk[O]) =
-          if (c.size < n) s -> c
-          else {
-            val (unit, rest) = c.splitAt(n)
-            resize(rest, s >> Pull.output1(unit))
-          }
+    fs2.Stream.eval(Semaphore[F2](n.toLong)).flatMap { demand =>
+      fs2.Stream.eval(Semaphore[F2](0L)).flatMap { supply =>
+        fs2.Stream
+          .eval(
+            Ref[F2].of(JunctionBuffer[O](Vector.empty[O], endOfSupply = None, endOfDemand = None))
+          )
+          .flatMap { buffer =>
```

Comment:
We sometimes use a `Stream.force` as a way of reducing code nesting.

```suggestion
    fs2.Stream.force { 
      for {
        demand <- Semaphore[F2](n.toLong)
        supply   <- Semaphore[F2](0L)
        buffer    <- Ref[F2].of(JunctionBuffer[O](Vector.empty[O], endOfSupply = None, endOfDemand = None))
      } yield { 
         /* here follows the stream definition*/
      }
```